codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/endpoints/EndpointResolverGenerator.java

```diff
@@ -0,0 +1,514 @@
+                                w.write("$L := *$L", getLocalVarParameterName(p), getMemberParameterName(p));
+                            });
+                        },
+                        "rules", generateRulesList(ruleset.getRules(), scope)));
+    }
+
+    private GoWriter.Writable generateEmptyResolveMethodBody() {
+        return goTemplate("return endpoint, $fmtErrorf:T(\"no endpoint rules defined\")", commonCodegenArgs);
+    }
+
+    private GoWriter.Writable generateResolverTypeDocs() {
+        return goDocTemplate("$resolverImplementationType:T provides the implementation for resolving endpoints.",
+                commonCodegenArgs);
+    }
+
+    private GoWriter.Writable generateResolveEndpointMethodDocs() {
+        return goDocTemplate("$resolveEndpointMethodName:L attempts to resolve the endpoint with the provided options,"
+                + " returning the endpoint if found. Otherwise an error is returned.",
+                commonCodegenArgs);
+    }
+
+    private GoWriter.Writable generateValidateParams(Parameters parameters) {
+        if (!haveRequiredParameters(parameters)) {
+            return emptyGoTemplate();
+        }
+        return goTemplate("""
+                if err = $paramArgName:L.$paramsValidateMethod:L(); err != nil {
+                        return endpoint, $fmtErrorf:T("endpoint parameters are not valid, %w", err)
+                    }
+                """,
+                commonCodegenArgs,
+                MapUtils.of(
+                        "paramsValidateMethod", VALIDATE_REQUIRED_FUNC_NAME));
+
+    }
+
+    private GoWriter.Writable generateRulesList(List<Rule> rules, Scope scope) {
+        return (w) -> {
+            rules.forEach(rule -> {
+                rule.getDocumentation().ifPresent(w::writeDocs);
+                w.write("$W", generateRule(rule, rule.getConditions(), scope));
+            });
+
+            if (!rules.isEmpty() && !(rules.get(rules.size() - 1).getConditions().isEmpty())) {
+                // TODO better error including parameters that were used?
+                w.writeGoTemplate("return endpoint, $fmtErrorf:T("
+                        + "\"no rules matched these parameters. This is a bug, %#v\", $paramArgName:L)",
+                        commonCodegenArgs);
+            }
+        };
+    }
+
+    private GoWriter.Writable generateRule(Rule rule, List<Condition> conditions, Scope scope) {
+        if (conditions.isEmpty()) {
+            return rule.accept(new RuleVisitor(scope, this.fnProvider));
+        }
+
+        var condition = conditions.get(0);
+        var remainingConditions = conditions.subList(1, conditions.size());
+
+        var generator = new ExpressionGenerator(scope, this.fnProvider);
+        var fn = conditionalFunc(condition);
+
+        String conditionName;
+        if (condition.getResult().isPresent()) {
+            var ident = condition.getResult().get();
+            conditionName = "_" + ident.asString();
```

Comment:
**nit**: `conditionName` is slightly confusing here, maybe `conditionIdentifier` since we are creating a variable/identifier?